freeipa/src/main/java/com/sequenceiq/freeipa/service/freeipa/user/UserKeytabService.java

```diff
@@ -0,0 +1,72 @@
+import com.sequenceiq.cloudbreak.auth.altus.Crn;
+import com.sequenceiq.cloudbreak.auth.altus.GrpcUmsClient;
+import com.sequenceiq.cloudbreak.auth.security.InternalCrnBuilder;
+import com.sequenceiq.cloudbreak.logger.MDCUtils;
+import com.sequenceiq.freeipa.controller.exception.BadRequestException;
+import com.sequenceiq.freeipa.kerberos.KerberosConfig;
+import com.sequenceiq.freeipa.kerberos.KerberosConfigRepository;
+import com.sequenceiq.freeipa.service.freeipa.user.kerberos.UserKeytabGenerator;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Service;
+
+import javax.inject.Inject;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static com.sequenceiq.freeipa.controller.exception.NotFoundException.notFound;
+
+@Service
+public class UserKeytabService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(UserKeytabService.class);
+
+    private static final String IAM_INTERNAL_ACTOR_CRN = new InternalCrnBuilder(Crn.Service.IAM).getInternalCrnForServiceAsString();
+
+    @Inject
+    private KerberosConfigRepository kerberosConfigRepository;
+
+    @Inject
+    private GrpcUmsClient grpcUmsClient;
+
+    @Inject
+    private UserKeytabGenerator userKeytabGenerator;
+
+    private String getKerberosRealm(String accountId, String environmentCrn) {
+        KerberosConfig krbConfig =  kerberosConfigRepository
+                .findByAccountIdAndEnvironmentCrnAndClusterNameIsNull(accountId, environmentCrn)
+                .orElseThrow(notFound("KerberosConfig for environment", environmentCrn));
+        return krbConfig.getRealm();
+    }
+
+    private void validateSameAccount(String userAccountId, String environmentCrn) {
+        String environmentCrnAccountId = Objects.requireNonNull(Crn.fromString(environmentCrn)).getAccountId();
+        if (!environmentCrnAccountId.equals(userAccountId)) {
+            throw new BadRequestException("User and Environment must be in the same account");
+        }
+    }
+
+    public String getKeytabBase64(String userCrn, String environmentCrn) {
+        String userAccountId = Objects.requireNonNull(Crn.fromString(userCrn)).getAccountId();
+        validateSameAccount(userAccountId, environmentCrn);
+
+        String realm = getKerberosRealm(userAccountId, environmentCrn);
+
+        GetActorWorkloadCredentialsResponse getActorWorkloadCredentialsResponse =
+                grpcUmsClient.getActorWorkloadCredentials(IAM_INTERNAL_ACTOR_CRN, userCrn, MDCUtils.getRequestId());
+        String workloadUsername = getActorWorkloadCredentialsResponse.getWorkloadUsername();
+        List<ActorKerberosKey> actorKerberosKeys = getActorWorkloadCredentialsResponse.getKerberosKeysList();
+
+        try {
+            return userKeytabGenerator.generateKeytabBase64(workloadUsername, realm, actorKerberosKeys);
+        } catch (IOException e) {
+            throw new RuntimeException("Failed to generate keytab", e);
```

Comment:
can you throw a more meaningful exception? perhaps make a custom exception so that we can map it to a response code in the api layer

Reply:
I've moved the exception handling into UserKeytabGenerator. I don't see how the IOException will ever trigger since we're passing our own ByteArrayOutputStream rather than doing file i/o. So I don't think a custom exception here is warranted.